### PR TITLE
[DNM] Azure Pipelines Local Pool Worker Discussion

### PR DESCRIPTION
**Pull request by Lnaden**
** DO NOT MERGE ** Experimental code which will not live inside the repo itself.

I've been experimenting with setting up a local worker for Azure DevOps Pipelines where we can install and do CI testing on code we have access to but cannot put online.

How its setup is what I have been tinkering with, but I wanted to discuss what we want it to do exactly once it is set up, hence the PR with the code. Since we will be running local hardware, it was suggested by @dgasmith to not merge this file in with the main repository and keep the file locked behind the Azure control panel.

The actions taken by the script are as follows:

1. Assign a path for a test conda env to be created which the Agent will destroy after each job best it can (won't show up on `conda info --env` either)
2. Show build information 
3. Create the (temp) environment and install QCEngine
3. Display post-install information
4. Run Pytest

Pipeline looks like this when run:

image

And the report from the Pytest block (cut blocks for space)
```
##[section]Starting: pytest
==============================================================================
Task         : Command line
Description  : Run a command line script using Bash on Linux and macOS and cmd.exe on Windows
Version      : 2.151.1
Author       : Microsoft Corporation
Help         : https://docs.microsoft.com/azure/devops/pipelines/tasks/utility/command-line
==============================================================================
Generating script.
========================== Starting Command Output ===========================
[command]/bin/bash --noprofile --norc /Users/levinaden/sandbox/azagent/osx/_work/_temp/efead106-138e-4efc-90b7-948561d743d2.sh
============================= test session starts ==============================
platform darwin -- Python 3.7.3, pytest-5.0.1, py-1.8.0, pluggy-0.12.0 -- /Users/levinaden/sandbox/azagent/osx/_work/_temp/testnv3.7/bin/python
cachedir: .pytest_cache
rootdir: /Users/levinaden/sandbox/azagent/osx/_work/1/s, inifile: setup.cfg
plugins: cov-2.7.1
collecting ... collected 144 items / 3 skipped / 141 selected

qcengine/programs/tests/test_dftd3_mp2d.py::test_dftd3_task[b3lyp-d3] PASSED [  0%]
...
qcengine/programs/tests/test_programs.py::test_psi4_task PASSED          [ 70%]
qcengine/programs/tests/test_programs.py::test_psi4_internal_failure PASSED [ 70%]
qcengine/programs/tests/test_programs.py::test_psi4_ref_switch PASSED    [ 71%]
qcengine/programs/tests/test_programs.py::test_rdkit_task SKIPPED        [ 72%]
qcengine/programs/tests/test_programs.py::test_rdkit_connectivity_error SKIPPED [ 72%]
qcengine/programs/tests/test_programs.py::test_torchani_task SKIPPED     [ 73%]
...
qcengine/tests/test_program_utils.py::test_check_program_avail[psi4] PASSED [ 86%]
qcengine/tests/test_program_utils.py::test_check_program_avail[torchani] SKIPPED [ 87%]
qcengine/tests/test_program_utils.py::test_check_program_avail[rdkit] SKIPPED [ 88%]
qcengine/tests/test_program_utils.py::test_program_avail_bounce PASSED   [ 88%]
...

=========================== short test summary info ============================
SKIPPED [1] /Users/levinaden/sandbox/azagent/osx/_work/1/s/qcengine/programs/tests/test_entos.py:10: could not import 'qcenginerecords': No module named 'qcenginerecords'
SKIPPED [1] /Users/levinaden/sandbox/azagent/osx/_work/1/s/qcengine/programs/tests/test_molpro.py:13: could not import 'qcenginerecords': No module named 'qcenginerecords'
SKIPPED [1] /Users/levinaden/sandbox/azagent/osx/_work/1/s/qcengine/programs/tests/test_terachem.py:11: could not import 'qcenginerecords': No module named 'qcenginerecords'
SKIPPED [20] /Users/levinaden/sandbox/azagent/osx/_work/1/s/qcengine/programs/tests/test_dftd3_mp2d.py:482: Psi4 requires at least Psi4 v1.3rc2
SKIPPED [20] /Users/levinaden/sandbox/azagent/osx/_work/1/s/qcengine/programs/tests/test_dftd3_mp2d.py:460: Psi4 requires at least Psi4 v1.3rc2
SKIPPED [1] qcengine/programs/tests/test_dftd3_mp2d.py:719: Not detecting module qcdb. Install package if necessary to enable tests.
SKIPPED [18] qcengine/programs/tests/test_dftd3_mp2d.py:741: Not detecting module mp2d. Install package if necessary to enable tests.
SKIPPED [1] qcengine/programs/tests/test_programs.py:96: Not detecting module rdkit. Install package if necessary to enable tests.
SKIPPED [1] qcengine/programs/tests/test_programs.py:109: Not detecting module rdkit. Install package if necessary to enable tests.
SKIPPED [1] qcengine/programs/tests/test_programs.py:126: Not detecting module torchani. Install package if necessary to enable tests.
SKIPPED [1] qcengine/tests/test_procedures.py:78: Not detecting module rdkit. Install package if necessary to enable tests.
SKIPPED [1] qcengine/tests/test_procedures.py:94: Not detecting module rdkit. Install package if necessary to enable tests.
SKIPPED [1] qcengine/tests/test_procedures.py:110: Not detecting module torchani. Install package if necessary to enable tests.
SKIPPED [1] qcengine/tests/test_program_utils.py:17: Not detecting module torchani. Install package if necessary to enable tests.
SKIPPED [1] qcengine/tests/test_program_utils.py:17: Not detecting module rdkit. Install package if necessary to enable tests.
SKIPPED [1] /Users/levinaden/sandbox/azagent/osx/_work/1/s/qcengine/tests/test_standard_suite.py:24: Program 'rdkit' not found.
SKIPPED [1] /Users/levinaden/sandbox/azagent/osx/_work/1/s/qcengine/tests/test_standard_suite.py:24: Program 'torchani' not found.
SKIPPED [1] /Users/levinaden/sandbox/azagent/osx/_work/1/s/qcengine/tests/test_standard_suite.py:37: Program 'rdkit' not found.
SKIPPED [1] /Users/levinaden/sandbox/azagent/osx/_work/1/s/qcengine/tests/test_standard_suite.py:37: Program 'torchani' not found.
SKIPPED [1] /Users/levinaden/sandbox/azagent/osx/_work/1/s/qcengine/tests/test_standard_suite.py:55: Program 'rdkit' not found.
SKIPPED [1] /Users/levinaden/sandbox/azagent/osx/_work/1/s/qcengine/tests/test_standard_suite.py:55: Program 'torchani' not found.
==================== 71 passed, 76 skipped in 17.86 seconds ====================
##[section]Finishing: pytest
```

### CODE REVIEW

File: .azure-pipelines.yml [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+        # Create test environment for package
+        conda env create -p $ENVPREFIX python=$(python.version) -f devtools/conda-envs/psi.yaml
+        conda activate $ENVPREFIX
+        # Build and install package
+        python setup.py develop --no-deps
+      displayName: "Conda Env Creation"
+
+    # Display MongoDB, Conda Environment, Python version
+    - script: |
+        sleep 5
```
Comment (dgasmith) [inline .azure-pipelines.yml]: No need for sleep anymore?

File: .azure-pipelines.yml [hunk 2]
```diff
@@ -0,0 +1,57 @@
+variables:
+  CONDASOURCE: "~/miniconda3/etc/profile.d/conda.sh"
```
Comment (dgasmith) [inline .azure-pipelines.yml]: Might want to make a user with poor permissions.

File: .azure-pipelines.yml [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+        free -m
+        df -h
+        ulimit -a
+      displayName: "Build Info"
+
+    # Setup python environment
+    - script: |
+        source $(CONDASOURCE)
+        # Create test environment for package
+        conda env create -p $ENVPREFIX python=$(python.version) -f devtools/conda-envs/psi.yaml
```
Comment (dgasmith) [inline .azure-pipelines.yml]: Ah great, so not pure Bash? I guess that has some upsides and downsides.

**Comment by dgasmith** [comment]
Ok, once we get the ARC VM lets go for it.

File: .azure-pipelines.yml (hunk 1 above)
Comment (Lnaden) [reply .azure-pipelines.yml]: Probably not, since this is now its own script.

File: .azure-pipelines.yml (hunk 2 above)
Comment (Lnaden) [reply .azure-pipelines.yml]: Agreed, its just using my local user on my laptop for testing. But its worth noting so I don't forget

File: .azure-pipelines.yml (hunk 3 above)
Comment (Lnaden) [reply .azure-pipelines.yml]: Its not pure bash, its able to interface with the local environment like bash, but I can also interact with the Azure Worker variables which are all parsed on script creation I think. Might not be the case, but I can access both of them all the same.

**Comment by Lnaden** [comment]
I think we have a plan then, I'm going to close this worker

**Pull request closed without merging**